# special-pdx
An experimental version of SMAP 


# SMAP split doc

SMAP should be a set of pieces of code to deal with sequencing data for xenograft.
SMAP split is to be used on a BAM file that was generated by mapping simulateneously on two genomes. SMAP split will generate two BAM files which corresponds to reads mapped on each of the genomes.

## 1. How to use it


For local (cit) usage, you should be sure your install supports it. On sv51 you have to run this line in you terminal before using it :
```source /datacit/Development/RNAseq_YR/DEVSMAP/smapenv/bin/activate```

Basic usage :
```python /path/to/script/smap_splitBySpecie_standard_0.1.py -i inputBamFile.bam```

Additional arguments :

	*	-g <graphSpecieID> : This is the prefix used to identify the specie of the graft sample. default : hs
	*	-h <hostSpecieID> :  This is the prefix used to identify the specie of the host sample. default : mmu
	*	-o <outputBamFilesPrefix> : This will be used to  name the resulting outputbam, one per specie. default is the name of the input bam
	*	-m : outputs multiple alignements
	*	-f <alignment scpre field in bam> : This should indicate in what additional field of the BAM file the alignement score is stored. Default is AS, which should work for both bwa and star
	*	-k : When this 	argument is given, the sorted bam file is kept. By default it is deleted.
	*	-v : Verbose mode, displays info on the run.
	*	â€”help : displays a help message


**Warning 1** : SMAP split will only work on a bam file that was generated by a mapping on a fasta file that contains the genomes of both studied species and for which the names of the chromosomes were modified, by adding a specie-specific prefix,to identify the specie of each of them. 

**Warning 2** : Current version of SMAP split will only work with a mapper that returns its alignment score as an additional bam field but also if it returns multiple alignments per reads.

**Warning 3** : Current version of SMAP split will only work on paired-end.


## 2. How it Works

SMAP split uses the pysam library to read, modify, filter and write reads from a bam file. It also borrows a piece of code from the htseq python package to process bundle of reads with the same names.


First, the input bam file is sorted by read name in order to get all alignments of the same reads in line.
Then, two output bam files are created, one for the host and the other for the graft. These bam files contain modified version of the original header containing only the chromosomes of the correponding specie and the prefix are stripped out from the chromosome  names.


The processing of a bundle of alignments of a given pair of reads (a framgent) begins :

1.	As some mappers (e.g. bwa) may not store the sequence or the sequence quality for all the alignments of a givrn read, both are saved before processing for read  1 and for read 2
2.	Two lists of alignment pairs are created, a list of true pairs and a list of asymetrical pair :
  *	truepairs (using function areTurePairs) : two alignments are considered as true pairs if :
  *	both are mapped
  *	both are paired
  *	both have same names
	*	one is read 1 and other is read 2
	*	reference and start of one is equal to reference and start of the pair of the other (both reference each other as their aligned pair, at least it terms of position)
	*	asymetricpairs (using function areAsymetricPairs) : two alignments are considered as asymetric pairs if :
  * both are mapped
  *	both have same names
  * one is read 1 and other is read 2
  * at least one reference and start of one is equal to reference and start of the pair of the other (at least one references each other as their aligned pair, at least it terms of position)
  * Note : true pairs of alignments are also asymetric pairs.

3.	If these lists are empty, the paired read is considred unaligned and it stops here for this fragment.
4.	The alignement score of a pair is computed as the sum of alignment score of each of the read alignment composing the pair. For asymetricPairs, the maximum score is computed and all the pairs with this score is considered as the best pairs. The same is done with truePairs, resulting in the maximum truepair score and the best true pairs.
5.	The reads of the best asymetric pairs are added into a set which removes duplicate alignments.
6.	For each read alignments (not pair) among the best asymetric pair set, the program looks if there is at least one mapping to the host genome (on a chromsome with a host prefix) and to the graft genome (on  a chromsome with a graft prefix)
7.	If there is both a mapping to the host and to the graft, the entire fragment is considered as ambiguous. The process stops here for this fragment
8.	If the -m option is given, all reads in the set of best asymetrical pairs are written to the host or graft bam depending on the outcome of (6), Otherwise, only the true pairs are written to the bam
9.	For each alignment to write, if the sequence or quality of the sequence is missing, it is added as stored in (1)
10.	Each alignment is written to the bam correspoonding to its specie and the index of its reference sequence (chromosome) is modified to correespond to the new header of the bam file.

